static-instrumenter/agent-instrumenter/src/main/java/io/opentelemetry/contrib/statical/instrumenter/ClassArchive.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.contrib.statical.instrumenter;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Enumeration;
+import java.util.Map;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipException;
+import java.util.zip.ZipOutputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Represents an archive storing classes (JAR, WAR). */
+class ClassArchive {
+
+  interface Factory {
+    default ClassArchive createFor(JarFile source, Map<String, byte[]> instrumentedClasses) {
+      return new ClassArchive(source, instrumentedClasses);
+    }
+  }
+
+  private static final Logger logger = LoggerFactory.getLogger(ClassArchive.class);
+
+  private final JarFile source;
+  private final Map<String, byte[]> instrumentedClasses;
+
+  ClassArchive(JarFile source, Map<String, byte[]> instrumentedClasses) {
+    this.source = source;
+    this.instrumentedClasses = instrumentedClasses;
+  }
+
+  void copyAllClassesTo(ZipOutputStream outJar) throws IOException {
+
+    Enumeration<JarEntry> inEntries = source.entries();
+    while (inEntries.hasMoreElements()) {
+      copyEntry(inEntries.nextElement(), outJar);
+    }
+  }
+
+  private void copyEntry(JarEntry inEntry, ZipOutputStream outJar) throws IOException {
+    String inEntryName = inEntry.getName();
+    ZipEntry outEntry =
+        inEntryName.endsWith(".jar") ? new ZipEntry(inEntry) : new ZipEntry(inEntryName);
+
+    try (InputStream entryInputStream = getInputStreamForEntry(inEntry, outEntry)) {
+      try {
+        outJar.putNextEntry(outEntry);
+        StreamUtils.copy(entryInputStream, outJar);
+        outJar.closeEntry();
+      } catch (ZipException e) {
+        if (!isEntryDuplicate(e)) {
+          logger.error("Error while creating entry: " + outEntry.getName(), e);
+          throw e;
+        }
+      }
+    }
+  }
+
+  private static boolean isEntryDuplicate(ZipException ze) {
+    return (ze.getMessage() != null && ze.getMessage().contains("duplicate"));
+  }
+
+  private InputStream getInputStreamForEntry(JarEntry inEntry, ZipEntry outEntry)
+      throws IOException {
+
+    InputStream entryIn = null;
+    ArchiveEntry entry = ArchiveEntry.ofFilePath(inEntry.getName());
+    if (entry.isClass() && !entry.shouldBeSkipped()) {
+      String className = entry.getClassName();
+      try {
+        // sanity check
+        Class.forName(className, false, ClassLoader.getSystemClassLoader());
```

Comment:
I wouldn't expect an archive being processed to have its classes on the system classpath. We should load classes into a separate classloader doing the transformation, or some dependencies of this tool could conflict with the transformed archive. We're probably not that far in this PR, but I would remove this sanity check in anticipation of that

Reply:
The basic idea was to run the tool in both java agent space (incoming part) and main space (this part) with all of the archives we need to instrument on a classpath. I can remove the sanity although initially it was helpful in debugging the M2 plugin (checking what exactly didn't end up on the instruemntors class path).

Reply:
I think it would still be better to have the archives needed to instrument on its own classloader, not the system one, to ensure no conflicts with the tool. I don't think that would affect the operation of teh javaagent

Reply:
Sanity removed.